### ISSUE TITLE
[REVIEW]: multimatch_gaze: The MultiMatch algorithm for gaze path comparison in Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1525]
**Submitting author:** @AdinaWagner (<a href="http://orcid.org/0000-0003-2917-3450">Adina Wagner</a>)
**Repository:** <a href="https://github.com/adswa/multimatch_gaze" target ="_blank">https://github.com/adswa/multimatch_gaze</a>
**Version:** v0.1.1
**Editor:** @kyleniemeyer
**Reviewer:** @FelixHenninger, @JonathanReardon
**Archive:** <a href="https://doi.org/10.5281/zenodo.3369531" target="_blank">10.5281/zenodo.3369531</a>

## Status

[![status](http://joss.theoj.org/papers/fb69f61ef3d07b7080bd1af6d579e270/status.svg)](http://joss.theoj.org/papers/fb69f61ef3d07b7080bd1af6d579e270)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/fb69f61ef3d07b7080bd1af6d579e270"><img src="http://joss.theoj.org/papers/fb69f61ef3d07b7080bd1af6d579e270/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/fb69f61ef3d07b7080bd1af6d579e270/status.svg)](http://joss.theoj.org/papers/fb69f61ef3d07b7080bd1af6d579e270)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/adswa/multimatch_gaze" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@FelixHenninger & @JonathanReardon, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @kyleniemeyer know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/adswa/multimatch_gaze">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.1.1)?
- [x] **Authorship:** Has the submitting author (@AdinaWagner) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/adswa/multimatch_gaze">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.1.1)?
- [x] **Authorship:** Has the submitting author (@AdinaWagner) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @FelixHenninger, @JonathanReardon it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by kyleniemeyer** [comment]
Hi @adswa @FelixHenninger @JonathanReardon 👋  the actual review will take place in here

**Comment by JonathanReardon** [comment]
Hello, (1) could the list of dependencies be added to the README.md? It states that all will be installed with pip, but it might be better to explicitly state what those dependencies are. (2) Could the formatting of the top (author/affiliation etc.) section of paper.md be improved for clarity?

**Comment by JonathanReardon** [comment]
The documentation is very good, and the "Example computation" is particularly useful (https://multimatch.readthedocs.io/en/latest/example.html). Do you state anywhere what you used to overlay the data onto movie frame image (Figure 1 and 2)? if not, I think it would be a useful information to add.

**Comment by adswa** [comment]
Hey @JonathanReardon,
Thanks for these helpful remarks.

The dependencies are now explicitly stated in the README and in the docs.

As for formatting the paper.md header section I'm unfortunately unsure how to do that, at the moment. @kyleniemeyer, is it possible to line-break those affiliations or similar?

Indeed, the "how to contribute" section was only a minor section in the README.md, and I have now added a CONTRIBUTING.md file (with guidelines on contributing, seeking help, and a code of conduct). 

I have linked the script used to overlay gaze onto the stimulus content (its publicly available as part of the studyforrest publication) in the example in the docs - I agree that this is helpful information. Also, many thanks for reading the documentation so attentively!

The changes made so far are in [this PR](https://github.com/adswa/multimatch_gaze/pull/30). I'll try to also add a header improvement to it, and will make a note here when I merge them into master.

**Comment by JonathanReardon** [comment]
Hi @adswa -- thank you for the swift response and the changes! I have now finished my part of the review (all boxes ticked). I thoroughly checked the tool and I found it very easy to use, it works exactly as stated. Docs/instructions were also thorough and very clear, which I very much appreciate. Note: I tested it on Linux Mint (19.1) -- I'm not sure what systems you've tested it on but it might be useful to know that it works on this distribution too. It's great to see another useful tool re-implemented for the open source world, congrats (to all involved)!

I'm not sure how it works with respect to the other reviewer, but from my end, I'm happy to approve for publication.

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by kyleniemeyer** [comment]
@adswa @JonathanReardon if you mean the affiliation info in the PDF linked in the last comment, the format is part of the JOSS template. The Markdown file is just the original source, so don't worry about how things look in that.

(also, one comment for @adswa: could you add more detail to the affiliation info for your colleague Yaroslav? That one should have the same info, like department, city, state, country.)

**Comment by JonathanReardon** [comment]
@kyleniemeyer Yes -- I did mean the markdown file, but of course how raw md looks isn't very important! Sorry @adswa, that may have come across as incredibly picky (you can ignore that comment), the reality is, I forgot what I was looking at for a second :) thanks all!

**Comment by adswa** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by adswa** [comment]
Thanks @kyleniemeyer, the affiliation is updated, and I merged all changes in [this commit](https://github.com/adswa/multimatch_gaze/commit/29636ffd0b0d76e45d08970b7652b4a8b4aab14c). No worries, @JonathanReardon! It did not come across in a bad way at all!

Have a good weekend everyone!

**Comment by kyleniemeyer** [comment]
Thanks for your feedback @JonathanReardon!

**Comment by kyleniemeyer** [comment]
@FelixHenninger have you had a chance to start your review?

**Comment by adswa** [comment]
Are there any updates yet, @kyleniemeyer? I just want to make sure I haven't missed a notification.

**Comment by kyleniemeyer** [comment]
@adswa we are just waiting on @FelixHenninger's review now; I was in touch via email a few days ago and got a confirmation that it should be completed soon.

**Comment by adswa** [comment]
Thanks a lot!

**Comment by kyleniemeyer** [comment]
Hi @FelixHenninger, just wanted to check in our your review—thanks!

**Comment by FelixHenninger** [comment]
Hej everyone, and sincerest apologies for the delay, which has been due to me being out of office in in July, followed by health issues on my part over the past ten days -- I'm very sorry for the trouble! My brief notes follow:

This project has been a genuine pleasure to review as the installation and use work flawlessly, and as a result I have only minor points to add to @JonathanReardon's excellent (and already-addressed) comments above. I emphatically agree that the documentation is fantastic, and also found the references describing the package's inner workings very clear and helpful when looking through the functionality. I had a couple of things that were unclear to me at first, but in every case it turned out that I hadn't read the docs carefully enough. 🙈

I found the manuscript very helpful and accessible, with a clear motivation and concise description of the package. I have only two minor formatting-related points that might be worth a second look:

* On lines 54/55, I think the citations might work better in the [in-text format](https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax) rather than in brackets.
* In the bibliography, I think that the PDF generation might have garbled the capitalization for Behavior Research Methods. There are also some minor inconsistencies with the DOI formatting, and some of the links don't work as a result (Burmeister & Mast, Dewhurst et al. and Johansson et al.).

There is one minor point I encountered that I think is the scope of the JOSS review, so please consider this as suggestion only (if they aren't covered elsewhere already -- I'd be happy to make an issue out of it if that's a better place for the information):
* It's really great to see CI testing in a project like this! 😍 One of the things I had missed initially is the automated testing, and thus I had tried running the test suite locally, which isn't (as far as I can see) documented -- it might be helpful for future contributors to briefly note how to trigger a test run manually.

Ok, I hope this makes sense -- again, thanks for putting together this super-neat package, and doing it with such care and visible attention to detail. If you, @adswa, could take another look at the manuscript formatting, I think this is a go from my perspective. 👍 Sorry again for the delay!

**Comment by adswa** [comment]
Hi @FelixHenninger, sorry to hear about health issues - don't worry about the delay. Glad to hear you find it useful, and thanks for your comments. :slightly_smiling_face: 

In this [PR](https://github.com/adswa/multimatch_gaze/pull/33) I've added a note about how to run the test suite locally into the README.md, CONTRIBUTING.md, and the docs. I find it to be a very useful addition, thanks for the suggestion!
I also adjusted the manuscript so that the citations you mentioned are now in-text, and fixed the broken dois (thanks for checking so thoroughly). I forgot the BRM capitalization and did this [here](https://github.com/adswa/multimatch_gaze/pull/34) :smile:

**Comment by adswa** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by adswa** [comment]
Retrying in order to get the doi with an underscore to work.

**Comment by adswa** [comment]
@whedon  generate pdf

**Comment by adswa** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by FelixHenninger** [comment]
Looks great to me, thanks @adswa for the lightning-fast changes! 👍 🎉

**Comment by kyleniemeyer** [comment]
Thanks for your review @FelixHenninger!

@adswa at this point I'm going to do a final review of your paper—we're in the home stretch!

**Comment by kyleniemeyer** [comment]
@adswa my only editorial request for the paper is to add
> A summary describing the high-level functionality and purpose of the software for a diverse, non-specialist audience.

(This is from our documentation on [what papers should contain](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain).)

This could take multiple forms, but the simplest is probably to add a sentence or two explaining the application to a general reader at the very beginning.

**Comment by adswa** [comment]
Thanks a lot, @kyleniemeyer, I've included a short description at the very beginning into the introduction in [this PR](https://github.com/adswa/multimatch_gaze/pull/37):

> ``multimatch-gaze`` is a Python package for computing the
similarity of scan paths from eye gaze data on five different dimensions.

Let me know if that suffices at as a high-level functionality description, or if that is not detailed enough. Thanks!

**Comment by kyleniemeyer** [comment]
@adswa this is better, but I think it still needs a bit more general explanation for non-experts—for example, I don't really know what scan paths are or what they are used for. Could you add a sentence explaining that?

**Comment by adswa** [comment]
Thanks a lot for the feedback, @kyleniemeyer! I made another attempt in [this PR](https://github.com/adswa/multimatch_gaze/pull/38)

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by kyleniemeyer** [comment]
@adswa that looks great! I just made a minor edit in this PR, if you could merge: https://github.com/adswa/multimatch_gaze/pull/39

Once that is complete, please archive the entire repository (e.g., using Zenodo) and let me know the DOI here.

**Comment by adswa** [comment]
All done, it is 10.5281/zenodo.3369531.

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01525/joss.01525/10.21105.joss.01525.pdf)

**Comment by kyleniemeyer** [comment]
@adswa sorry, one last thing—can you edit the title of the Zenodo archive to match the title of the paper?

**Comment by adswa** [comment]
oh sure! sorry, I wasn't aware. Should be done now.

**Comment by kyleniemeyer** [comment]
@whedon set 10.5281/zenodo.3369531 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3369531" target="_blank">10.5281/zenodo.3369531</a> is the archive.

**Comment by kyleniemeyer** [comment]
@openjournals/joss-eics ok, this submission is ready to accept!

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/905

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/905, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
👋 @kyleniemeyer - I think @openjournals/joss-eics members can do this themselves, even if not on duty - at least @arfon and I have in the past. There's certainly no requirement to, but it's an option if you want to.

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1162338558201475077) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/906
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01525](https://doi.org/10.21105/joss.01525)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks to @FelixHenninger and @JonathanReardon for reviewing and @kyleniemeyer for editing!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01525/status.svg)](https://doi.org/10.21105/joss.01525)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01525">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01525/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01525/status.svg
   :target: https://doi.org/10.21105/joss.01525
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01525/status.svg)](https://doi.org/10.21105/joss.01525)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss

**Comment by adswa** [comment]
Thanks a lot to everyone involved! This has been a great experience! :tada:

**Comment by FelixHenninger** [comment]
Congrats @adswa for the awesome work, I'm glad to have helped a little with this!

**Comment by JonathanReardon** [comment]
I'm very happy to have helped out on this too, it was a pleasure, congrats! and thank you all!
